#259 Validation for breached passwords

DragonBe opened the pull request:
This is the UndisclosedPassword validator discussed in #259 which uses PSR-18 http client interface for communicating with [Have I Been Pwned?](https://www.haveibeenpwned.com) web service for validating passwords in a secure, safe way using K-Anonymity method.

**IMPORTANT:** The thing that might be a problem for this pull request is that PSR-18 composer package requires PHP 7.0 or higher.

All functionality is documented and provided with unit tests ensuring 100% code coverage.
![image](https://user-images.githubusercontent.com/282825/56085067-9ce85a80-5e3d-11e9-99e3-bd2a8745369e.png)

Please review this pull request and let me know if I need to change or improve the functionality.

DragonBe committed:
Adding psr/http-client as suggested package

To make this validator PSR18 compliant we are implementing the `psr/http-client` interfaces.

DragonBe committed:
Writing out the validator for undisclosed passwords

We can now validate if a password has been seen in a password breach.

DragonBe committed:
Fixing coding style violations

Running CodeSniffer indicated some issues

DragonBe committed:
Creating a HttpClientException for testing purposes

Mocking an interface implementing \Throwable was not possible, creating a dummy \Exception class implementing the interface is the easiest solution without complexity requiring further testing.

DragonBe committed:
Removing unreached exceptions

The input is verified at an earlier stage so no need for throwing an exception for a condition that never will be reached.

DragonBe committed:
Improving tests

Adding coverage and more precise testing we can now achieve higher quality with less maintenance.

DragonBe committed:
Implementing PSR-17 for making requests

Instead of recreating a new request object, I'm using the request factory to build a new request object.

This was indirectly suggested by @azjezz via Twitter (https://twitter.com/azjezz/status/1117112058032656384).

One thing is for sure: I ❤️ the PHP community for they give me so much wisdom and make me a smarter person 🙂

Ocramius commented on composer.json:
```diff
@@ -30,7 +30,9 @@
         "zendframework/zend-session": "^2.8",
-        "zendframework/zend-uri": "^2.5"
+        "zendframework/zend-uri": "^2.5",
+        "psr/http-client": "^1.0",
```
These should be near the other `psr/*` libs above

DragonBe replied:
@Ocramius I can do that… but have you seen my request in regards PSR-18 not being compatible with PHP 5.6? ➡️ https://github.com/zendframework/zend-validator/issues/259#issuecomment-482839690

Ocramius replied:
Yeah, I think we need to bump the dependency here: it is long overdue and should just be done.

Also, I'm using this occasion to re-express to @zendframework/community-review-team the fact that bumping PHP version can be freely done in minor releases.

Ocramius replied:
I raised https://github.com/zendframework/maintainers/issues/59 about this

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+    const UNKNOWN_ERROR = 'unknownError';
+
+    protected $messageTemplates = [
+        self::PASSWORD_BREACHED => 'The provided password was used by others',
```
Reading this in the UI can be confusing for a user: who is "others" to them?

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+
+    protected $messageTemplates = [
+        self::PASSWORD_BREACHED => 'The provided password was used by others',
+        self::WRONG_INPUT => 'The provided password failed to be correctly hashed, please verify your input',
```
Given usages of `WRONG_INPUT`, I'd say that:

 * the constant should be renamed to `NOT_A_STRING` (or something like that)
 * the message should reflect the same

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+    protected $messageTemplates = [
+        self::PASSWORD_BREACHED => 'The provided password was used by others',
+        self::WRONG_INPUT => 'The provided password failed to be correctly hashed, please verify your input',
+        self::CONNECTION_FAILURE => 'Unable to reach HIBP service, please try again later',
```
HIBP is a text that doesn't make sense for the typical end-user: we'll need a better error message here too

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+        self::PASSWORD_BREACHED => 'The provided password was used by others',
+        self::WRONG_INPUT => 'The provided password failed to be correctly hashed, please verify your input',
+        self::CONNECTION_FAILURE => 'Unable to reach HIBP service, please try again later',
+        self::UNKNOWN_ERROR => 'Something happened beyond our control, error reporting should give more details',
```
Also here: this error message is developer-friendly, but not end-user friendly

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+    /**
+     * @var int
+     */
+    private $count = 0;
```
What count is this representing? Looking just at the property, this can be confusing

DragonBe replied:
Count is used for providing a "count" of amount of times a password was found in Have I been pwned. Each hash that returns contains a number. See https://haveibeenpwned.com/API/v2#SearchingPwnedPasswordsByRange for details.

Ocramius replied:
> amount of times a password was found

`pwnedPasswordsFoundCount`?

DragonBe replied:
Or I could implement `Countable` so you could count on the object directly.

Ocramius replied:
I don't think that would clarify it: it would make it even more confusing instead. Why would a `Countable` implemented on a `ValidatorInterface` implementation returning a "count of pwned found passwords"? This is just about communicating what a specific `private` state is meant for, so we shouldn't expand the `public` API of the object for an implementation detail that is already ambiguous in `private` scope.

DragonBe replied:
Good point, in a "validation" context it has no real purpose. I'll remove it.

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+    /**
+     * @var RequestFactoryInterface
+     */
+    private $httpRequest;
```
s/httpRequest/makeHttpRequest maybe? (since this isn't a real request, but the factory for it)

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+    /**
+     * @var ResponseFactoryInterface
+     */
+    private $httpResponse;
```
s/httpResponse/makeHttpResponse maybe? (since this isn't a real response, but the factory for it)

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+     */
+    private $count = 0;
+
+    /**
```
Docblock can be omitted, as it doesn't contain further/useful information

DragonBe replied:
Even though not providing useful information beyond the default value being provided, IDE's tend to enjoy this and PHPDocumentor still uses it to generate good API documentation from it.

I would like to keep it in if you don't mind.

Ocramius replied:
API documentation and IDE document this from the type declaration. If the docblock does not provide information about a symbol that isn't already part of the reflection of said symbol, it is noise. I don't have a hard stance on this, but in `doctrine/coding-standard`, this sort of docblock is explicitly forbidden, for example.

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+        } catch (Exception\RuntimeException $runtimeException) {
+            $this->error(self::CONNECTION_FAILURE);
+            return false;
+        } catch (\Exception $exception) {
```
I don't think we should catch `\Exception` here: something went really wrong here, maybe in service wiring, so we should let the exception bubble up instead, making developers aware of the problem

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+
+        try {
+            $response = $this->httpClient->sendRequest($request);
+        } catch (ClientExceptionInterface $connectException) {
```
You are catching/re-throwing and then catching in the caller again here: if we can recognise precise `ClientExceptionInterface` types, let's react to them in the caller, and otherwise let exceptions bubble up, since otherwise misconfiguration scenarios become extremely hard to inspect.

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+        return true;
+    }
+
+    private function isPwnedPassword($password)
```
Noticed while looking at the lack of type declarations: new code should target `develop` (target branch needs changing in the PR)

DragonBe replied:
Moved base to `develop` but even there the composer refers to php 5.6 so until your proposal is agreed upon, let's leave out the type declarations for now. It's a jiffy to put them in, but I'm already struggling with a failed pass on PHP 5.6 because of PSR18 and I'd like to tackle one issue at the time. Let's solve zendframework/maintainers#59 first before adding type declarations.

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+            $this->error(self::UNKNOWN_ERROR);
+            return false;
+        }
+        if ($isPwnd) {
```
This variable is redundant - should be removed

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+     */
+    private function getRangeHash($passwordHash)
+    {
+        $range = substr($passwordHash, self::HIBP_RANGE_BASE, self::HIBP_RANGE_LENGTH);
```
Variable redundant

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+    const HIBP_API_TIMEOUT = 300;
+    const HIBP_CLIENT_UA = 'zend-validator';
+    const HIBP_CLIENT_ACCEPT = 'application/vnd.haveibeenpwned.v2+json';
+    const HIBP_RANGE_LENGTH = 5;
```
Looking at usages of the constants below, this makes kinda sense, but the naming of the constants themselves do not always make sense. If we can expand the names, that would be best.

Ocramius commented on src/UndisclosedPassword.php:
```diff
@@ -0,0 +1,177 @@
+
+final class UndisclosedPassword extends AbstractValidator
+{
+    const HIBP_API_URI = 'https://api.pwnedpasswords.com';
```
@weierophinney all of these constants should be `private` in order to prevent needless BC breaks later on: can we please bump to `^7.1` before expanding the API surface beyond what should be `public`?

Ocramius commented on test/UndisclosedPasswordTest.php:
```diff
@@ -0,0 +1,233 @@
+     *
+     * @param string $password
+     *
+     * @covers \Zend\Validator\UndisclosedPassword::__construct
```
Consider using a single `@covers \Zend\Validator\UndisclosedPassword` at class level: this sort of fine-grained coverage is more added maintenance work when someone splits a `private` detail into more `private` methods

DragonBe committed:
Moving PSR components in composer.json

As suggested by @Ocramius in PR #264 I'm moving added PSR components to the existing PSR requirement in composer.json

DragonBe committed:
Processing feedback from PR #264

Applying the following changes:

- Improving error messages in the validator
- Improving naming conventions for variable names and constants
- Bubbling up important Exceptions to be taken care of higher up the stream
- Removing unused or arbitrary variables, messages and constants
- Simplifying coverage declarations in unit tests

DragonBe committed:
Removing password found count

After reviewing the feedback provided by @Ocramius on PR #264 I've agreed with him that keeping the count of found passwords has no meaning in a validation context and have removed it from the code base.

DragonBe committed:
Adding documentation for UndisclosedPassword validator

No feature is complete if there's not some user documentation available, so with this I've provided a basic introduction on how to use the validator.

froschdesign commented on doc/book/validators/undisclosed-password.md:
````diff
@@ -0,0 +1,38 @@
+If the password was found via the service, `isValid` will return FALSE. If the password was not found, `isValid` will return TRUE.
+
+```php
+$validator = new Zend\Validator\UndisclosedPassword(
````
The example must be completely reproducible for the end user. At the moment, the example shows the same thing as the API, which is not very helpful. We need a full code example including a HTTP client and the factories.

DragonBe replied:
I will update this piece of example code after I figure out which PSR-18 http client I could use in this example. 

I was thinking about using the Guzzle6 adapter and factories, because zend-diactoros doesn't implement PSR-18 (yet), but I can go any direction.

DragonBe replied:
@froschdesign In commit 1383af3 I have provided a working example of `Zend\Validator\UndisclosedPassword` using PSR7, PSR17 and PSR18 HTTP Client and Messages.

Please have a look at it if this is in line with your expectations. If not, please indicate how I can improve it.

froschdesign commented on doc/book/validators/undisclosed-password.md:
```diff
@@ -0,0 +1,38 @@
+
+To validate if a password was disclosed in a known data breach, you need to provide a HTTP Client that implements `psr\http\ClientInterface`, a `Psr\Http\Message\RequestFactoryInterface` and a `Psr\Http\Message\ResponseFactoryInterface` to the constructor and validate the password you want to check.
+
+If the password was found via the service, `isValid` will return FALSE. If the password was not found, `isValid` will return TRUE.
```
Also set "false" and "true" in backticks and use lowercase letters. Thanks!

froschdesign reviewed (changes requested):
First: 👍 for adding documentation.
Second: I added some small comments. The important part is the code example because we need a full working example.

DragonBe committed:
Processing minor feedback points in UndisclosedPassword documentation

I'm processing the easy to fix documentation issues pointed out by @froschdesign in PR #264

DragonBe committed:
Creating a full qualified URI

Since we have no knowledge how the client is configured, I'm using a fully qualified URI pointing to the HIBP API service. This will prevent issues making a connection.

DragonBe committed:
Providing a working example documentation

To make it simple for users to make use of this validation feature, I created a working example people can use.

weierophinney committed:
fixup: move and edit docs

weierophinney committed:
qa: make constants private

Adds license docblock to new class files.

Makes constants in `UndisclosedPassword` private.

Updates `UndisclosedPasswordTest` to use `ReflectionClass` when doing
constant comparisons.

weierophinney committed:
feat: bump PHP requirement to 7.1, and support 7.4

Bumps the minimum supported PHP version to 7.1, to allow us to use
private constants (and reduce BC breaks going forwards).

Adds support for PHP 7.4 to the test matrix.

weierophinney committed:
fix: runs update_hostname_validator to update hostname list

And hopefully fix Travis test errors.

weierophinney committed:
fix: add typehints to UndisclosedPassword

On all internal methods.

weierophinney committed:
qa: remove intermediate variable

weierophinney committed:
docs: adds CHANGELOG entry for #264

weierophinney commented:
Thanks, @DragonBe! With the imminent transition to the Laminas Project, I am expecting we will likely revisit the ability to bump a PHP version during a major series, and, as such, decided to do so here, as it will ensure we do not have any BC issues with your new class in the future. I added visibility to all constants, and added typehints to parameters and return values of all internal methods. Will release shortly with version 2.13.0.
